src/backend/distributed/planner/function_call_delegation.c

```diff
@@ -0,0 +1,288 @@
+ *    has metadata.
+ *
+ * Copyright (c), Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "catalog/pg_proc.h"
+#include "catalog/pg_type.h"
+#include "commands/defrem.h"
+#include "distributed/citus_custom_scan.h"
+#include "distributed/citus_ruleutils.h"
+#include "distributed/colocation_utils.h"
+#include "distributed/commands.h"
+#include "distributed/commands/multi_copy.h"
+#include "distributed/connection_management.h"
+#include "distributed/function_call_delegation.h"
+#include "distributed/master_metadata_utility.h"
+#include "distributed/master_protocol.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_executor.h"
+#include "distributed/multi_physical_planner.h"
+#include "distributed/remote_commands.h"
+#include "distributed/shard_pruning.h"
+#include "distributed/version_compat.h"
+#include "distributed/worker_manager.h"
+#include "nodes/makefuncs.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/parsenodes.h"
+#include "nodes/primnodes.h"
+#include "optimizer/clauses.h"
+#include "parser/parse_coerce.h"
+#if PG_VERSION_NUM >= 120000
+#include "parser/parsetree.h"
+#endif
+#include "miscadmin.h"
+#include "tcop/dest.h"
+#include "utils/lsyscache.h"
+#include "utils/syscache.h"
+
+static bool contain_param_walker(Node *node, void *context);
+
+/*
+ * contain_param_walker scans node for Param nodes.
+ * returns whether any such nodes found.
+ */
+static bool
+contain_param_walker(Node *node, void *context)
+{
+	return IsA(node, Param);
+}
+
+
+/*
+ * TryToDelegateFunctionCall calls a function on the worker if possible.
+ * We only support delegating the SELECT func(...) form for distributed
+ * functions colocated by distributed tables, and not more complicated
+ * forms involving multiple function calls, FROM clauses, WHERE clauses,
+ * ... Those complex forms are handled in the coordinator.
+ */
+DistributedPlan *
+TryToDelegateFunctionCall(Query *query)
+{
+	FromExpr *joinTree = NULL;
+	List *targetList = NIL;
+	TargetEntry *targetEntry = NULL;
+	FuncExpr *funcExpr = NULL;
+	DistObjectCacheEntry *procedure = NULL;
+	Oid colocatedRelationId = InvalidOid;
+	Const *partitionValue = NULL;
+	Datum partitionValueDatum = 0;
+	ShardInterval *shardInterval = NULL;
+	List *placementList = NIL;
+	DistTableCacheEntry *distTable = NULL;
+	Var *partitionColumn = NULL;
+	ShardPlacement *placement = NULL;
+	WorkerNode *workerNode = NULL;
+	StringInfo queryString = NULL;
+	Task *task = NULL;
+	Job *job = NULL;
+	DistributedPlan *distributedPlan = NULL;
+	int32 groupId = 0;
+
+	if (!CitusHasBeenLoaded() || !CheckCitusVersion(DEBUG4))
+	{
+		/* Citus is not ready to determine whether function is distributed */
+		return NULL;
+	}
+
+	groupId = GetLocalGroupId();
+	if (groupId != 0 || groupId == GROUP_ID_UPGRADING)
+	{
+		/* do not delegate from workers, or while upgrading */
```

Comment:
We don't have to do anything right now, but why we do not delegate from workers? That should help in MX case. As I said, it is OK to not do it now, just wanted to learn if are there any specific concerns you have?

Reply:
Delegation loops. Benefit is also unclear because connections are a scarce resource in mx.